src/attestation/cose_sign.ts

```diff
@@ -0,0 +1,158 @@
+      this.protectedHeader = protectedHeader;        // Protected header
+      this.unprotected = unprotected;          // Unprotected header
+      this.payload = payload;                  // Payload
+      this.signature = signature;              // Signature
+    }
+    static async decodeCBOR(data: Uint8Array) {
+      return await cbor.decodeFirst(data);
+    }
+  
+    // Get protected header as a map
+    async getProtectedHeader() {
+        try {
+            return await COSE_Sign1.decodeCBOR(this.protectedHeader);
+        } catch (err) {
+            throw new Error('SerializationError : Failed to decode protected header');
+        }
+    }
+    
+    private getAllCerts(cert : typeof X509Certificate ,cabundle : Buffer[]){
+      let allCerts: typeof X509Certificate[] = [cert];
+
+      // Iterate over cabundle and process each buffer to create X509 certificates
+      for (const certBuffer of cabundle) {
+        try {
+          // Check if certBuffer is actually a Buffer, and then process it
+          if (Buffer.isBuffer(certBuffer)) {
+            const cert = new X509Certificate(certBuffer);  // Create X509Certificate from buffer
+            allCerts.push(cert);  // Add certificate to the allCerts array
+          } else {
+            throw new Error("Invalid certificate format. Expected Buffer.");
+          }
+        } catch (error) {
+          throw new Error(`Failed to parse certificate: ${error}`);
+        }
+      }
+      return allCerts;
+    }
+
+    // Core verification function
+    public async verifySignature(key : SigningPublicKey) {
+        // Retrieve algorithm and digest from the key
+        // const { algorithm, digest } = key.getParameters();
+
+        // Decode protected header
+        const protectedHeaderMap = await this.getProtectedHeader();
+  
+        // Extract and validate the signature algorithm from the protected header
+        const protectedAlgVal = protectedHeaderMap.get(1);
+        if (!protectedAlgVal) {
+            throw new Error(
+                'SpecificationError : Protected Header does not contain a valid Signature Algorithm specification'
+            );
+        }
+
+        if (protectedAlgVal !== -35) {
+            // Signature algorithm mismatch
+            return false;
+        }
+  
+        // Construct the signature structure based on protected header and payload
+        // const sigStructure = this.createSigStructure();
+        const sigStructure = [
+          'Signature1',
+          this.protectedHeader,
+          EMPTY_BUFFER,
+          this.payload
+        ]
+        const ToBeSigned = cbor.encode(sigStructure);
+        // const byteArray = Buffer.from(sigStructure.toString('hex'), 'hex');
+        // Compute the digest
+        const hash = createHash('sha384');
+        hash.update(ToBeSigned);
+        const structDigest = hash.digest();
+
+        const sigs = { r: this.signature.slice(0, 48), s: this.signature.slice(48) };
```

Comment:
use signature.length/2 instead of 48 since it's not guaranted that the length will be 96